org.eclipse.jdt.core/dom/org/eclipse/jdt/core/dom/CompilationUnitResolverDiscovery.java

```diff
@@ -0,0 +1,68 @@
+/*******************************************************************************
+ * Copyright (c) 2024 Red Hat, Inc. and others.
+ *
+ * This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License 2.0
+ * which accompanies this distribution, and is available at
+ * https://www.eclipse.org/legal/epl-2.0/
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *******************************************************************************/
+package org.eclipse.jdt.core.dom;
+
+import org.eclipse.core.runtime.CoreException;
+import org.eclipse.core.runtime.IConfigurationElement;
+import org.eclipse.core.runtime.IExtension;
+import org.eclipse.core.runtime.IExtensionPoint;
+import org.eclipse.core.runtime.ILog;
+import org.eclipse.core.runtime.Platform;
+import org.eclipse.jdt.core.JavaCore;
+import org.eclipse.jdt.internal.core.dom.ICompilationUnitResolver;
+
+class CompilationUnitResolverDiscovery {
+	private static final String SELECTED_SYSPROP = "ICompilationUnitResolver"; //$NON-NLS-1$
+	private static final String COMPILATION_UNIT_RESOLVER_EXTPOINT_ID = "compilationUnitResolver" ; //$NON-NLS-1$
+	private static boolean ERROR_LOGGED = false;
+
+
+	static ICompilationUnitResolver getInstance() {
+		String compilationUnitResolverId = System.getProperty(SELECTED_SYSPROP);
+		IExtensionPoint extension = Platform.getExtensionRegistry().getExtensionPoint(JavaCore.PLUGIN_ID, COMPILATION_UNIT_RESOLVER_EXTPOINT_ID);
+		if (extension != null && compilationUnitResolverId != null && !compilationUnitResolverId.isEmpty()) {
+			IExtension[] extensions = extension.getExtensions();
+			for (IExtension ext : extensions) {
```

Comment:
The system property can change during the lifecycle of the IDE, so we may need to consider a different extension when this happens. And we do really prefer instantiating 1 new object one every query instead of using a singleton as this can allow to keep state information in the class.
I'm going to make a comment about it.

Note that the usual pattern is 1 reason why we have to restart Eclipse after installing/uninstalling plugins. If all consumers were re-reading the extension points on demand (or better if they were reacting to changes in extension registry), then we wouldn't have to restart the IDE for changes to take effect.
I'll try to implement a cache so that we keep the "current" extension until the system property change.